### DISCUSSION TITLE
Import cockpit.js while developing plugins

### DISCUSSION THREAD

**Discussion by jbbjarnason** [cockpit-project/cockpit#19364] [Q&A] [answered]
I would like to develop plugins without running them in full blown cockpit environment(to use debuggers and hot reloading) and I am trying to authenticate to the websocket server. 

A minimal example:

```js
import logo from './logo.svg';
import './App.css';
import cockpit from "./cockpit";

function App() {

  const newDbus = window.cockpit.dbus('com.skaginn3x.ipc_ruler', { superuser: 'try' });
  const newDbusObjectProxy = newDbus.proxy('com.skaginn3x.manager', '/com/skaginn3x/ipc_ruler');
  console.log('useEffect', window.cockpit);
  console.log('com.skaginn3x.ipc_ruler', newDbusObjectProxy);
  console.log('valid', newDbus.valid);
  console.log('proxy valid', newDbusObjectProxy.valid);

  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}

export default App;
```

I guess I would need to make a login call before creating the dbus connection, I have tried to use curl but I am not sure how I would use it,

```bash
curl -X GET -u <user>:<pass> http://localhost:9090/cockpit/login
{"csrf-token":"<token>"}
```

For reference I am using this doc: https://github.com/cockpit-project/cockpit/blob/main/doc/authentication.md

So the question is, how can I import a local copy of cockpit.js and authenticate correctly via the websocket?

**Comment by martinpitt** [comment]
What is a "full blown cockpit environment"? You certainly need to use `cockpit-ws` to serve the pages and provide the websocket, not a custom webserver (like build tools that do hot reloading). But it doesn't need to be that "hot", as long as you just work on HTML/JS pages, you don't  have to log out and back into your session, you can just force-reload the page to use new code.

> how can I import a local copy of cockpit.js

`import cockpit` is quite alright, provided that you import cockpit's pkg/lib, like e.g. https://github.com/cockpit-project/starter-kit/ does. Or copy the file and put it into your own tree. Then it will be bundled into your page.

Alternatively, you can also use the file shipped by cockpit-bridge package:

```html
<script src="../base1/cockpit.js"></script>
```
There's not much difference between the two, unless you want to make/test some local modifications to cockpit.js of course.

> I guess I would need to make a login call before creating the dbus connection

Pages don't have to do that -- cockpit-ws handles authentication, and then sets a cookie, which the pages and cockpit.js use to talk to the websocket.

**Reply by jbbjarnason** [reply]
> You certainly need to use cockpit-ws to serve the pages and provide the websocket

This is what I would like to bypass.
I would like to serve the plugin with npm, but connect to the cockpit-ws via the `cockpit.js` here  https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/cockpit.js#L476  with hardcoded ws url, https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/cockpit.js#L315

When I connect to the cockpit-ws in such circumstances the login procedure has not been done.

So is there a way for me to do a login call and get the authentication cookie and somehow inject it to `cockpit.js` ?

**Reply by martinpitt** [reply, marked as answer]
Sorry, this is not supported. cockpit-ws does way more than just serving plain files (which is a bit annoying, and we want to clean it up at some point, but that's what it is today)

**Reply by martinpitt** [reply]
Note: The "magic" is **not** in cockpit.js -- that's more or less just syntactic sugar for the [wire protocol](https://github.com/cockpit-project/cockpit/blob/main/doc/protocol.md), and it's really just a library. The magic is in cockpit-ws.
